File: easy/1221-Split-a-String-in-Balanced-Strings.c
// Returns the maximum number of balanced strings that can be obatined
int balancedStringSplit(char *s){
    int out = 0; // Tracks the current number of balanced strings
    char *pointer = s; // Pointer to the current character
    int ls = 0; // Tracks the number of 'L's in the current substring
    int rs = 0; // Tracks the number of 'R's in the current substring

    while(*pointer != '\0') // While the pointer is on the string
    {
        if(*pointer == 'L')
        {
            ls++;
        }
        if(*pointer == 'R')
        {
            rs++;
        }

        if(ls == rs) // If the current substring is balanced
        {
            out++; // Increment the output
            ls = 0; // Reset the values
            rs = 0;

        }

        pointer++; // Move the pointer to the next character
    }

    return out;
}

File: easy/1221-Split-a-String-in-Balanced-Strings.hs
recur :: String -> Int -> Int -> Int -> Int
-- Base case -> return number of balanced strings (o)
recur "" l r o = if (l == 0 && r == 0) || l /= r then o
                 else o+1
-- Recursive case -> increment correct counter
recur s l r o = if (l == 0 && r == 0) || l /= r then
                   if (s !! 0) == 'L' then recur (tail s) (l+1) r o
                   else recur (tail s) l (r+1) o
                else
                   recur s 0 0 o+1

-- Returns the maximum number of balanced strings that can be obatined
balancedStringSplit :: String -> Int
balancedStringSplit s = recur s 0 0 0